### ISSUE TITLE
Bad modes in the PM solver.

### ISSUE THREAD

**Issue by rainwoodman** [rainwoodman/MP-Gadget#208]
There appears to be an abnormal growth (slower) near the domain edge when checking the power spectrum of the PM solver.

The effect is more evident in smaller runs where the domain edge is in the large scale. In bluetides it is shifted to the far right that we didn't bother too much about it.

The road map to fix this:

1. reenable the PM dumps; probably taking the code that writes fourier mesh into bigfile from fastpm.
2. run with 4 ranks
3. run with 1 ranks
4. load them to python and look for artifacts along domain edges, and along tree node edges.
5. In the first case, look for bugs in the CIC painting routine; for artifacts of the second kind, look for tree node syncrhonization errors. The second error is unlikely now that we rebuild tree every step.

**Comment by rainwoodman** [comment]
@sbird Could you check if the DM power at 'early' redshifts also suffer the same hump, similar to the baryons and neutrinos?

**Comment by sbird** [comment]
For neutrinos, it grows over time. It is too small too see at z=9, vaguely noticeable sometimes at z=4 if you squint and always visible at z=2. Then it gets larger at lower redshifts. Here are the neutrino powers for M_nu = 0.06 as a function of redshift:
[pks-lowmass_nu-0_2.pdf](https://github.com/MP-Gadget/MP-Gadget/files/2018220/pks-lowmass_nu-0_2.pdf)
[pks-lowmass_nu-0_3333.pdf](https://github.com/MP-Gadget/MP-Gadget/files/2018222/pks-lowmass_nu-0_3333.pdf)
[pks-lowmass_nu-0_5.pdf](https://github.com/MP-Gadget/MP-Gadget/files/2018221/pks-lowmass_nu-0_5.pdf)
[pks-lowmass_nu-1.pdf](https://github.com/MP-Gadget/MP-Gadget/files/2018224/pks-lowmass_nu-1.pdf)

Will check DM next.

**Comment by sbird** [comment]
I see no such feature in the DM. Is it possible I have an off-by-one error in the neutrino code?

[pks-lowmasss0_2.pdf](https://github.com/MP-Gadget/MP-Gadget/files/2018245/pks-lowmasss0_2.pdf)
[pks-lowmasss0_5.pdf](https://github.com/MP-Gadget/MP-Gadget/files/2018246/pks-lowmasss0_5.pdf)
[pks-lowmasss0_3333.pdf](https://github.com/MP-Gadget/MP-Gadget/files/2018247/pks-lowmasss0_3333.pdf)
[pks-lowmasss1.pdf](https://github.com/MP-Gadget/MP-Gadget/files/2018248/pks-lowmasss1.pdf)

**Comment by sbird** [comment]
Ah, if I plot the saved delta_m (ie, the total matter power that is passed to the neutrino integrator), then I do see this feature in the right place (note the k values are unitless and linear in this plot): It appears first fairly early on and then just grows. I bet there is a zero mode somewhere in the PM code.
![deltas_lowmass](https://user-images.githubusercontent.com/1325948/40253219-d81c8170-5a93-11e8-839e-dbced44e63ff.png)

**Comment by sbird** [comment]
Ok, the feature is being hidden in the CDM power spectrum by the rebinning to make sure there are enough modes in each bin. It is present in the raw data.

**Comment by sbird** [comment]
I can check this by running a short simulation with a different initial realisation. We can also hide it in the paper plots by rebinning the neutrino power spectrum if we want to.

**Comment by sbird** [comment]
And this is a single point with exactly 8 modes in the bin. This is a low-mode fluctuation.

**Comment by rainwoodman** [comment]
Which bin is it? How many modes does are the bins around it have? This is probably one place where the linear binning worked better than log binning, where the number of modes is strictly increasing -- log binning after removing Nmodes=0 won't be like that.

**Comment by rainwoodman** [comment]
However, if it is cosmic variance then this mode shouldn't have been growing differently relative to the neighboring modes.  Do you happen to know the i,j,k of the 8 modes in that bin? Would it correspond to the 2 d decomposition scheme in anyway (the job should have reported the process mesh)?

**Comment by rainwoodman** [comment]
Do you by any chance have the baryon power spectrum?

**Comment by sbird** [comment]
So this is the 40th non-zero bin, bin 295, and the rows surrounding it are:
0.142049 0.618228 48 25325.4
0 0 0 0
0 0 0 0
0.145104 0.58108 8 23803.6
0 0 0 0
0.146608 0.575429 54 23572.1
0.148096 0.562038 84 23023.6

I do not know the i,j,k of the 8 modes, and there were no baryons in the simulation, just CDM and neutrinos. The Seed value was 9281110. and the IC generator was S-GenIC. Would you like me to run a simulation with a different seed to check?

It could be cosmic variance and still evidence for an inaccuracy in our solver, that it is amplifying cosmic variance somehow...

**Comment by sbird** [comment]
It is also present in the no-neutrino simulation in just the same place: so that rules out the neutrinos.
![nonu](https://user-images.githubusercontent.com/1325948/40263886-10479ac8-5ace-11e8-9d8c-a957e2e904de.png)

**Comment by rainwoodman** [comment]
The fundamental frequency is 2pi / 300, so this mode has ii + jj + kk = 48.
```
>> (0.142049 / (2 * pi / 300.)) ** 2, (0.145104 / (2 * pi / 300.)) ** 2, (0.146608 / (2 * pi / 300.)) ** 2

(46.00013795612613, 48.00003364954637, 49.000229166897626)
```
The only solution to 48 is the mode 4, 4, 4. The 8 modes are +-4, +-4, +-4.

The MP-GenIC scheme does not produce a gaussian with a negative feature at this frequency.
Rather it produced a huge spike at this frequency. 

![image](https://user-images.githubusercontent.com/138060/40264220-be344a1e-5ad3-11e8-9ab4-919499c047f2.png)

If it was non-linear evolution damping this mode, then it should affected dark matter too. 

Does number 4 happen to match with the domain decomposition of the FFT? Btw do you have this power and other modes
```
0.11847687835088977
0.1203138387230014
0.12566370614359174
0.13410672042894523
0.14510394913873742
0.15812336265110555
0.17270824457886919
0.18849555921538758
```

These are modes that have 4, 4 on x,y but varying z. If they all look weird then it must be cross the fft domains.

**Comment by sbird** [comment]
It is very possible that I have inverted the sign of the Gaussian in S-GenIC.

**Comment by sbird** [comment]
Let me attach a few power spectra to the bug report. I don't know the FFT decomposition for this job (the slurm log got accidentally deleted), but it is easy enough to confirm with a different number of processors.

**Comment by sbird** [comment]
A few power spectra from the DM-only run.

[powerspectrum-0.0100.txt](https://github.com/MP-Gadget/MP-Gadget/files/2019092/powerspectrum-0.0100.txt)
[powerspectrum-0.0109.txt](https://github.com/MP-Gadget/MP-Gadget/files/2019093/powerspectrum-0.0109.txt)
[powerspectrum-0.0200.txt](https://github.com/MP-Gadget/MP-Gadget/files/2019094/powerspectrum-0.0200.txt)
[powerspectrum-0.0404.txt](https://github.com/MP-Gadget/MP-Gadget/files/2019095/powerspectrum-0.0404.txt)
[powerspectrum-0.1000.txt](https://github.com/MP-Gadget/MP-Gadget/files/2019096/powerspectrum-0.1000.txt)

**Comment by sbird** [comment]
Here is the same plot with a different seed - there are equivalently sized bumps but now at different places. I will next make sure I get the same thing with a different number of processors.

![Uploading nonu-seed+1.png…]()

**Comment by sbird** [comment]
I also confirm that the power spectra are entirely unchanged if I alter the number of processors used.

**Comment by sbird** [comment]
and nothing looks particularly weird about those other modes, that I could see.

**Comment by rainwoodman** [comment]
Here is a comparison of the DM growth between you pasted and FastPM on the same seed, with unitary amplitude (since your spectra looked so)

![image](https://user-images.githubusercontent.com/138060/40324072-3827c98a-5cec-11e8-92e5-6d7755625fde.png)

There is no extra growth at this mode from FastPM. Something is still wrong here I'll check with a pure PM wrong with and without Tree force.

**Comment by sbird** [comment]
Yes, tree force may be it. Ack. Something rotten in the state of Denmark.

**Comment by rainwoodman** [comment]
No. Disabling the tree, running with 2x2 ranks still have similar features. 

![image](https://user-images.githubusercontent.com/138060/40329746-631db326-5cff-11e8-91d6-00076e752f1e.png)

Blue crosses were your file, green were mine (128**3 on 2x2) without tree force. vertical lines are 2 * k0 and 4 * k0. I am starting to suspect ZA ic could be causing this. Will check running with FastPM's 2LPT ic.

**Comment by sbird** [comment]
That is very interesting. Can we have a force test, ala one of the figures in the Gadget-2 paper? It has been on my todo list for a while, but I never got around to it.

**Comment by rainwoodman** [comment]
I think I figured why. The fastpm spectra I used were widely binned, not down to individual |k|.

Here, I ran fastpm and Gadget without Tree on the same time steps, using the standard fastpm kernel (ik / k**2 ish ) against Gadget kernel (four point / k**2, disabling smoothing and CIC compensation)

Using the very fine binning (down to individual mode vectors norms), both fastpm and MP-Gadget shows poor scattering . The scattering in FastPM is actually bigger. 

![image](https://user-images.githubusercontent.com/138060/40336979-b16645ee-5d21-11e8-81ca-bb5f1013f84e.png)

I think at this point we shall agree that both PM codes are giving these non-linear growth on individual modes. The gadget kernel (which uses low noise 4 point differentiation) appears to give a less noisy spectra.

**Comment by rainwoodman** [comment]
No something is not right. We see same number of orange points and purple points, suggesting the same bins are used.

**Comment by rainwoodman** [comment]
Here is an updated figure where the actual points sampled from fastpm power of wide bins are shown. The previous figure were with linear interpolation, leaving the impression there are same number of sample points in the orange series.
 
![image](https://user-images.githubusercontent.com/138060/40337186-ea37b442-5d22-11e8-9950-4f4b051b2d80.png)

**Comment by sbird** [comment]
Ok. I think this is pretty clearly mode scatter then. Shall we close the bug?
